In [5]:
import numpy as np
import pandas as pd

# Loading the dataset

In [6]:
df=pd.read_csv('../data/movies_metadata.csv')
df.head().transpose()

0  \
adult                                                              False   
backdrop_path                           /lxD5ak7BOoinRNehOCA85CQ8ubr.jpg   
belongs_to_collection  {'name': 'Toy Story Collection', 'poster_path'...   
budget                                                          30000000   
genres                 [{'name': 'Animation', 'id': 16}, {'name': 'Co...   
homepage                            http://toystory.disney.com/toy-story   
id                                                                   862   
imdb_id                                                        tt0114709   
original_language                                                     en   
original_title                                                 Toy Story   
overview               Led by Woody, Andy's toys live happily in his ...   
popularity                                                        52.706   
poster_path                             /uXDfjJbdP4ijW5hWSBrPrlKpxab.jpg   
production_companies   [{'name': 'Pixar', 'origin_country': 'US', 'id...   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-10-30   
revenue                                                        373554033   
runtime                                                               81   
spoken_languages                [{'iso_639_1': 'en', 'name': 'English'}]   
status                                                          Released   
tagline                                                              NaN   
title                                                          Toy Story   
video                                                              False   
vote_average                                                         7.9   
vote_count                                                         12976   

                                                                       1  \
adult                                                              False   
backdrop_path                            /y7SQmjlB42VvYyRIFQXLQ4ZYrn.jpg   
belongs_to_collection  {'name': 'Jumanji Collection', 'poster_path': ...   
budget                                                          65000000   
genres                 [{'name': 'Adventure', 'id': 12}, {'name': 'Fa...   
homepage                     http://www.sonypictures.com/movies/jumanji/   
id                                                                  8844   
imdb_id                                                        tt0113497   
original_language                                                     en   
original_title                                                   Jumanji   
overview               When siblings Judy and Peter discover an encha...   
popularity                                                         8.122   
poster_path                             /bdHG5Mo83VPobeZZdlSz0Y7HQHB.jpg   
production_companies   [{'name': 'TriStar Pictures', 'origin_country'...   
production_countries   [{'iso_3166_1': 'US', 'name': 'United States o...   
release_date                                                  1995-12-15   
revenue                                                        262797249   
runtime                                                              104   
spoken_languages       [{'iso_639_1': 'en', 'name': 'English'}, {'iso...   
status                                                          Released   
tagline                        Roll the dice and unleash the excitement!   
title                                                            Jumanji   
video                                                              False   
vote_average                                                         7.2   
vote_count                                                          7583   

                                                                       2  \
adult                                    

# understanding the dataset
The dataset above was obtained through the TMDB API. The movies available in this dataset are in correspondence with the movies that are listed in the MovieLens Latest Full Dataset comprising of 26 million ratings on 60,000 movies from 27,000 users. Let us have a look at the features that are available to us.

In [7]:
df.columns

Index(['adult', 'backdrop_path', 'belongs_to_collection', 'budget', 'genres',
       'homepage', 'id', 'imdb_id', 'original_language', 'original_title',
       'overview', 'popularity', 'poster_path', 'production_companies',
       'production_countries', 'release_date', 'revenue', 'runtime',
       'spoken_languages', 'status', 'tagline', 'title', 'video',
       'vote_average', 'vote_count'],
      dtype='object')

# features
- adult: Indicates if the movie is X-Rated or Adult.
- backdrop_path: Background image for the movie
- belongs_to_collection: A stringified dictionary that gives information on the movie series the particular film belongs to.
- budget: The budget of the movie in dollars.
- genres: A stringified list of dictionaries that list out all the genres associated with the movie.
- homepage: The Official Homepage of the move.
- id: The ID of the move.
- imdb_id: The IMDB ID of the movie.
- original_language: The language in which the movie was originally shot in.
- original_title: The original title of the movie.
- overview: A brief blurb of the movie.
- popularity: The Popularity Score assigned by TMDB.
- poster_path: The URL of the poster image.
- production_companies: A stringified list of production companies involved with the making of the movie.
- production_countries: A stringified list of countries where the movie was shot/produced in.
- release_date: Theatrical Release Date of the movie.
- revenue: The total revenue of the movie in dollars.
- runtime: The runtime of the movie in minutes.
- spoken_languages: A stringified list of spoken languages in the film.
- status: The status of the movie (Released, To Be Released, Announced, etc.)
- tagline: The tagline of the movie.
- title: The Official Title of the movie.
- video: Indicates if there is a video present of the movie with TMDB.
- vote_average: The average rating of the movie.
- vote_count: The number of votes by users, as counted by TMDB.

In [9]:
df.shape

(61768, 25)

In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61768 entries, 0 to 61767
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   adult                  61768 non-null  bool   
 1   backdrop_path          39298 non-null  object 
 2   belongs_to_collection  5967 non-null   object 
 3   budget                 61768 non-null  int64  
 4   genres                 61768 non-null  object 
 5   homepage               11259 non-null  object 
 6   id                     61768 non-null  int64  
 7   imdb_id                61750 non-null  object 
 8   original_language      61768 non-null  object 
 9   original_title         61768 non-null  object 
 10  overview               60867 non-null  object 
 11  popularity             61768 non-null  float64
 12  poster_path            55990 non-null  object 
 13  production_companies   61768 non-null  object 
 14  production_countries   61768 non-null  object 
 15  re

There are a total of <b>61,768 movies</b> with <b>25</b> features. Most of the features have very few NaN values (apart from <b>homepage, backdrop_path, belongs_to_collection and tagline</b>). We will attempt at cleaning this dataset to a form suitable for analysis.

# Data cleaning
The data that was originally obtained was in the form of a JSON File. This was converted manually into a CSV file to arrive at an input that could be loaded into a Pandas DataFrame effortlessly. In other words, the dataset we have in our hands is already relatively clean. We will however attempt at learning more about our features and performing appropriate wrangling steps to arrive at a form that is more suitable for analysis.

Let us start by removing the features that are not useful to us.

In [11]:
df=df.drop(['backdrop_path'],axis=1)

since, backdrop_path is just a image, we dont need it for our analysis

lets have a look at feature title, original_title

In [12]:
df.loc[df['original_title']!=df['title'],['title','original_title']].head()

title                    original_title
46  The City of Lost Children        La Cité des Enfants Perdus
47             Shanghai Triad                         摇啊摇，摇到外婆桥
50           Wings of Courage  Guillaumet, les ailes du courage
88  The City of Lost Children        La Cité des Enfants Perdus
89             Shanghai Triad                         摇啊摇，摇到外婆桥

The original title refers to the title of the movie in the native language in which the movie was shot. As such, I will prefer using the translated,we will drop the original titles feature. We will be able to deduce if the movie is a foreign language film by looking at the original_language feature so no tangible information is lost in doing so.

In [13]:
df=df.drop('original_title',axis=1)

In [15]:
df.revenue.value_counts()

0           52146
12000000       26
10000000       26
11000000       25
2000000        21
            ...  
45858563        1
1476356         1
16178959        1
11417362        1
48854305        1
Name: revenue, Length: 8695, dtype: int64

We see that the majority of the movies have a recorded revenue of 0. This indicates that we do not have information about the total revenue for these movies. Although this forms the majority of the movies available to us, we will still use revenue as an extremely important feature going forward from the remaining 7000 moves.

In [16]:
df['revenue'] = df['revenue'].replace(0, np.nan)

In [17]:
df['budget'].value_counts()

0           50579
10000000      379
5000000       325
2000000       297
3000000       280
            ...  
2515000         1
10920000        1
213084          1
12434           1
959300          1
Name: budget, Length: 1708, dtype: int64

In [18]:
df['budget'] = df['budget'].replace(0, np.nan)

In [28]:
df.budget.isnull().sum()

50579


As we move forward trying to answer certain questions, we will have to construct several features suitable for that particular query. For now, we will construct two very important features:

- <b>return:</b> The ratio of revenue to budget.
- <b>year: </b> The year in which the movie was released.

The return feature is extremely insightful as it will give us a more accurate picture of the financial success of a movie. Presently, our data will not be able to judge if a $200 million budget movie that earned $100 million did better than a $50,000 budget movie taking in $200,000. This feature will be able to capture that information.

A return value > 1 would indicate profit whereas a return value < 1 would indicate a loss.

In [33]:
df['return'] = df['revenue'] / df['budget']


In [48]:
df['return'].notnull().sum()

6769

We have close to 7000 movies for which we have data on revenue and budget ratio. This is close to 10% of the entire dataset. Although this may seem small, this is enough to perform very useful analysis and discover interesting insights about the world of movies.

In [49]:
df['year'] = pd.to_datetime(df['release_date'], errors='coerce').apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

In [50]:
df['adult'].value_counts()

False    61752
True        16
Name: adult, dtype: int64


There are close to 0% adult movies in this dataset. The adult feature therefore is not of much use to us and can be safely dropped.

In [51]:

df = df.drop('adult', axis=1)

In [54]:
# save the data
df.to_csv('../data/movies_metadata_cleaned.csv')

# cast and crew

Let us now take a look at the cast and crew of our movies. We do not have these details with us in our main dataset. However, we have a separate file consisting of the full cast and crew credits of all the Movielens Movies. Let us take a look at this credits data.

In [55]:
credits_df=pd.read_csv('../data/credits_metadata.csv')

In [56]:
credits_df.head()

cast  \
0  [{'cast_id': 14, 'character': 'Woody (voice)',...   
1  [{'cast_id': 1, 'character': 'Alan Parrish', '...   
2  [{'cast_id': 2, 'character': 'Max Goldman', 'c...   
3  [{'cast_id': 1, 'character': 'Savannah Jackson...   
4  [{'cast_id': 1, 'character': 'George Banks', '...   

                                                crew     id  
0  [{'name': 'Andrew Stanton', 'credit_id': '52fe...    862  
1  [{'name': 'Larry J. Franco', 'credit_id': '52f...   8844  
2  [{'name': 'Gary Frutkoff', 'credit_id': '5af99...  15602  
3  [{'name': 'Forest Whitaker', 'credit_id': '52f...  31357  
4  [{'name': 'Alan Silvestri', 'credit_id': '52fe...  11862

<b> Credits Dataset </b>
- <b>cast:</b> A stringified list of dictionaries consisting of cast names and the corresponding characters they played.
- <b>crew:</b> A stringified list of dictionaries consisting of crew names and the function they performed.
- <b>id:</b> The TMDB ID of movie

We need to perform a left join of our original movies metadata dataframe with the credits dataframe on the TMDB Movie ID. Before we are able to perform this join, we need to make sure that the ID column of our main dataframe is clean and of type integer. To do this, let us try to perform an integer conversion of our IDs and if an exception is raised,we will replace the ID with NaN. We will then proceed to drop these rows from our dataframe

In [63]:
def convert_int(x):
    try:
        return int(x)
    except:
        return np.nan

In [64]:
df['id'] = df['id'].apply(convert_int)

In [65]:
df[df['id'].isnull()]

Empty DataFrame
Columns: [belongs_to_collection, budget, genres, homepage, id, imdb_id, original_language, overview, popularity, poster_path, production_companies, production_countries, release_date, revenue, runtime, spoken_languages, status, tagline, title, video, vote_average, vote_count, return, year]
Index: []

[0 rows x 24 columns]

In [66]:
df['id'] = df['id'].astype('int')

In [67]:
df = df.merge(credits_df, on='id')
df.shape

(61838, 26)

In [68]:
df['cast'] = df['cast'].apply(ast.literal_eval)
df['crew'] = df['crew'].apply(ast.literal_eval)

In [69]:
df.cast.head()

0    [{'cast_id': 14, 'character': 'Woody (voice)',...
1    [{'cast_id': 14, 'character': 'Woody (voice)',...
2    [{'cast_id': 14, 'character': 'Woody (voice)',...
3    [{'cast_id': 14, 'character': 'Woody (voice)',...
4    [{'cast_id': 1, 'character': 'Alan Parrish', '...
Name: cast, dtype: object

In [70]:
df['cast_size'] = df['cast'].apply(lambda x: len(x))
df['crew_size'] = df['crew'].apply(lambda x: len(x))

In [71]:
df.cast_size.head()

0    40
1    40
2    40
3    40
4    32
Name: cast_size, dtype: int64

In [74]:
df['cast'] = df['cast'].apply(lambda x: [i['name'] for i in x] if isinstance(x, list) else [])

In [75]:
df.cast.head()

0    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...
1    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...
2    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...
3    [Tom Hanks, Tim Allen, Don Rickles, Jim Varney...
4    [Robin Williams, Jonathan Hyde, Kirsten Dunst,...
Name: cast, dtype: object

In [76]:
def get_director(x):
    for i in x:
        if i['job'] == 'Director':
            return i['name']
    return np.nan

In [77]:

df['director'] = df['crew'].apply(get_director)

In [78]:
df.director.head()

0    John Lasseter
1    John Lasseter
2    John Lasseter
3    John Lasseter
4     Joe Johnston
Name: director, dtype: object

In [81]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 61838 entries, 0 to 61837
Data columns (total 29 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   belongs_to_collection  5971 non-null   object 
 1   budget                 11207 non-null  float64
 2   genres                 61838 non-null  object 
 3   homepage               11277 non-null  object 
 4   id                     61838 non-null  int32  
 5   imdb_id                61820 non-null  object 
 6   original_language      61838 non-null  object 
 7   overview               60937 non-null  object 
 8   popularity             61838 non-null  float64
 9   poster_path            56056 non-null  object 
 10  production_companies   61838 non-null  object 
 11  production_countries   61838 non-null  object 
 12  release_date           61772 non-null  object 
 13  revenue                9640 non-null   float64
 14  runtime                61661 non-null  float64
 15  sp

In [90]:
s = df.apply(lambda x: pd.Series(x['cast']),axis=1).stack().reset_index(level=1, drop=True)
s

C:\Users\kiran\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  """Entry point for launching an IPython kernel.


0                 Tom Hanks
0                 Tim Allen
0               Don Rickles
0                Jim Varney
0             Wallace Shawn
                ...        
61837       Fernando Piazza
61837    Margaret Rose Keil
61837        Claudine Damon
61837          Vera Besusso
61837      Carlo Hintermann
Length: 936310, dtype: object

In [91]:
s.name='actor'
s

0                 Tom Hanks
0                 Tim Allen
0               Don Rickles
0                Jim Varney
0             Wallace Shawn
                ...        
61837       Fernando Piazza
61837    Margaret Rose Keil
61837        Claudine Damon
61837          Vera Besusso
61837      Carlo Hintermann
Name: actor, Length: 936310, dtype: object

In [92]:
cast_df = df.drop('cast', axis=1).join(s)

In [93]:
cast_df['actor'].head()

0        Tom Hanks
0        Tim Allen
0      Don Rickles
0       Jim Varney
0    Wallace Shawn
Name: actor, dtype: object

In [ ]:
cast_df.to_csv('../data/movies_metadata_castncrew.csv')